### ISSUE TITLE
Synchronous using closes connection just after sending a single message

### ISSUE THREAD

**Issue by nnseva** [CJWorkbench/channels_rabbitmq#13]
I would like to send messages from synchronous celery task.

I am sending a message using code like:

```python
from channels.layers import get_channel_layer
from asgiref.sync import async_to_sync
...
    async_to_sync(get_channel_layer().group_send)('simulation_event',{'type':'simulation_event','data':{...}})
```

As a result, I have a login warning on every call, it means that the connection is reestablished there:
```
In [6]: async_to_sync(get_channel_layer().group_send)('simulation_event', {'type':'simulation_event','data':{}})                                                                                                                             
only PLAIN login_method is supported, falling back to AMQPLAIN

In [7]: async_to_sync(get_channel_layer().group_send)('simulation_event', {'type':'simulation_event','data':{}})                                                                                                                             
only PLAIN login_method is supported, falling back to AMQPLAIN
```

Sequential calls to the get_channel_layer() return the same instance
```
In [15]: get_channel_layer()                                                                                                                                                                                                                 
Out[15]: <channels_rabbitmq.core.RabbitmqChannelLayer at 0x7f96c21cb9e8>

In [16]: get_channel_layer()                                                                                                                                                                                                                 
Out[16]: <channels_rabbitmq.core.RabbitmqChannelLayer at 0x7f96c21cb9e8>
```
so caching of the returned value will not help:
```
In [17]: l = get_channel_layer()                                                                                                                                                                                                             

In [18]: async_to_sync(l.group_send)('simulation_event', {'type':'simulation_event','data':{}})                                                                                                                                              
only PLAIN login_method is supported, falling back to AMQPLAIN

In [19]: async_to_sync(l.group_send)('simulation_event', {'type':'simulation_event','data':{}})                                                                                                                                              
only PLAIN login_method is supported, falling back to AMQPLAIN
```

looking to the code I've noticed that the connection is closed on the loop closing:
```python
    def _create_connection(self, loop):
    ...
        def _wrapper(self, *args, **kwargs):  # self = loop
            # If the event loop was closed, there's nothing we can do anymore.
            if not self.is_closed():
                self.run_until_complete(connection.close())
        
            with manager._connections_lock:
                if self in manager._connections:
                    del manager._connections[self]
    ...
```

It looks like the connection lifetime should be synchronized with a layer lifetime, not loop lifetime, just because the layer can live longer than the loop in this case.

**Comment by adamhooper** [comment]
Thanks for reporting this. We've seen it before in https://github.com/CJWorkbench/channels_rabbitmq/issues/4#issuecomment-484585816 ... but your phrasing is more exact. I'll explain this as clearly as I can; future bug reports can be marked as duplicates of this one. (You won't be the last person waylaid by Channels here.)

TL;DR An asyncio network connection can't outlast `loop.close()`. This is wontfix. You should restructure your code.

You've asked for connections to outlast the event loop; but that won't happen, and you're passing the buck. The _real_ question is: how can you avoid creating spurious event loops in your code? In production, there should be one event loop and it should never close.

The usual culprit in Django Channels is `async_to_sync()`. Here's how `async_to_sync()` is _meant_ to work:

1. The ASGI server (e.g., Daphne) creates an event loop on startup to handle connections.
2. Each incoming request gets handled in an async function (supplied by Django Channels) on the event loop.
3. To handle slow-running code (e.g., WSGI handlers such as Django views), the ASGI server calls `sync_to_async(wsgi_handler)()`. This runs `wsgi_handler()` in a thread of the event loop's default thread pool. But right before running `wsgi_handler()`, `async_to_sync()` _sets a thread-local variable, `main_event_loop`, pointing to the event loop_.
4. Now comes `async_to_sync()`. When you call `async_to_sync(async_function)()` _from within `wsgi_handler()`_, that schedules `async_function()` on the thread-local `main_event_loop`.

**Here's the problem**: if you call `async_to_sync()` from a thread that was _not_ managed by `sync_to_async()`, then the thread won't have a `main_event_loop`. This is an error 99% of the time ... but `async_to_sync()` won't complain. It will spin up a whole new event loop, run your function, and then shut down the event loop. You don't want this.

**Solution**: in your production app, _only ever_ call `async_to_sync()` from a thread that is managed by `sync_to_async()` -- such as a Django view. _Never_ call `async_to_sync()` from other threads.

That means **on other threads, `async_to_sync()` is the wrong call**. Instead, pass your event loop`loop` to those threads. Then they can use [asyncio.run_coroutine_threadsafe()](https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe). That returns a `Future`. If you need to wait until the future completes, you can call [future.result()](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result).

Here's a correct way to run async functions from non-`sync_to_async()` threads:

```python
# Assuming your event loop, `loop`, is in scope:
channel_layer = get_channel_layer()
asyncio.run_coroutine_threadsafe(channel_layer.group_send('my-group', {
    'type': 'script.event',
    'message': 'Hello World!',
}), loop).result()
```

**Comment by nnseva** [comment]
Hi @adamhooper,

Big thanx for help!

The most appropriate way I've found is to change the code to the following:

```python
    loop = asyncio.get_event_loop()
    loop.run_until_complete(get_channel_layer().group_send(my-group', {
        'type': 'script.event',
        'data': {...},
   }))
```
Code looking like this works exactly as I expected, not closing the connection and completing an asynchronous task before the return.

**Comment by adamhooper** [comment]
@nnseva Where does your event loop come from?

Your code looks dangerous. `loop.run_until_complete()` _stops the event loop_ once the future completes. That's a bad idea on production. The RabbitMQ connection may die if it does not send heartbeat messages to the server; and it can't send anything while the event loop is stopped.

On a production web server, never call `run_until_complete()` or `stop()` on an event loop.

**Comment by nnseva** [comment]
@adamhooper really I need to send some messages using channels in the absolutely synchronous code of the celery task which generates events from time to time. I think it would be one of the most popular cases for channels.

I understand, that the event loop is not evaluated between these calls. Yes, it may mean that the connection will be lost sometimes.

But. I hope that the backend will restore the connection automatically if it is lost. In fact, it should do it IMHO, avoiding message loss.

**Comment by adamhooper** [comment]
@nnseva It sounds like you need the event loop running forever, _and_ you need a blocking function running forever.

The solution is threads. You want an event loop on one thread and your blocking Celery-worker code on another.

Here's one approach: in your Celery worker app's `__main__` function, create an event loop. Schedule your regular `main()` within it: `loop.run_until_complete(sync_to_async(my_celery_main)())`. (You don't _need_ to use `sync_to_async()` here -- you could use `loop.run_in_executor(my_celery_main()); loop.run_forever()` instead -- but this one wrapping `sync_to_async()` would let you use `async_to_sync()` throughout your code without problems.)

You're right that this should be one of the most popular cases for Channels. I, too, find it startling that something so simple is so complex.

Celery 5.0 will support asyncio.

**Comment by nnseva** [comment]
@adamhooper hmmm ... such complexity may mean that `channels_rabbitmq` appears to be much less applicable in this case.

The `channels` package with the embedded `Redis` backend allows using the code as I've quoted in the topic without problems (except message loss in case of overload, which pushed me to search for your package). It means for me that the external plugin such a `channels_rabbitmq` should allow the same code.

It allows such using as a fact but makes some extra (connection) work in this case. Such extra work slows down productivity. Ok, I can try to avoid connection phase using `run_until_complete` instead of `async_to_sync`. You told me that such code can't be properly evaluated. It means that applying `channels_rabbitmq` to `channels` package leads to increasing code complexity to threading and other stuff which really makes my code more complex than if I just use pure `RabbitMQ` code. So, what the reason should be to continue using `channels_rabbitmq` package in this case?

From my point of view, such a problem should lead to searching such a solution which allows using `channels_rabbitmq` package without changes in the code, or with minimal changes.

The solution without changes at all should allow the way of using `async_to_sync` as recommended in `channels` documentation. It may mean that such a solution should avoid closing connection when the loop is closed, and support a connection life while the layer lives instead as I've told at the top.

The solution with minimal changes should recommend using something like `run_until_complete` or other `channels` or `asyncio` function instead of `async_to_sync`.

If connection life support requires additional threading, independently of solution chosen, your package probably should provide, or recommend, some additional function to be called at the start in case of synchronous code (which starts additional thread processing connections).

From my point of view, threading is not required always - if `run_until_complete` or `async_to_sync` is called right along with the whole program instance lifetime, the connection will be supported enough. Even the connection is broken, the `channels` backend should restore it: such behavior follows from the `channels` functionality.

Only such a solution makes changes minimal and simple enough to make using of `channels_rabbitmq`  justified for most programmers IMHO.

**Comment by adamhooper** [comment]
@nnseva You simply cannot expect stability if you keep stopping and starting your event loop. Nothing was designed to work that way. I'm telling you this and the Redis backend's authors aren't; but what I'm saying applies to the Redis backend as well. Don't shoot the messenger.

I agree that this is too complex. I deny all responsibility. I'm investing hours in finding decent coding patterns and pointing out broken ones.

**Comment by adamhooper** [comment]
I've read up a bit more about Celery. It appears that:

* Celery pre-5.0 can't interact with a library that uses asyncio
* Celery pre-5.0 can't interact with a library that uses threads
* Celery 4.x has an event loop; but it's a custom event loop (called "kombu"), incompatible with asyncio; channels_rabbitmq likely won't run on it.

@nnseva Your options are all distasteful:

1. Use `async_to_sync()`, reconnecting every time you send a message. This is what `channels_redis` does. (We can try and figure out how to nix the warning; that may help you forget how ugly this is.)
2. Replace `channels_rabbitmq` with a _synchronous_ AMQP client, and manage connections in your Celery-worker codebase. This sounds painful to me -- Celery is restrictive, so many libraries won't work. But the `channels_rabbitmq` data model is stable, so you can keep using `channels_rabbitmq` on your web server.
3. Replace Celery with an async-friendly task queue. This is what we do here at Workbench. (We prefer microservices to a monolithic worker.)
4. Wait for Celery 5.0, which should mesh with `channels_rabbitmq` swimmingly.
5. Use your `run_until_complete()` hack ... but that's a usage we just can't support at Workbench, because it may cause random-seeming problems in production.

**Comment by nnseva** [comment]
@adamhooper just for information.

I've tested a huge timeout (almost 20 minutes) between two calls of `run_until_complete` like I've stated above, for `channels` with `channels_rabbitmq` and `channels_redis` backends.

The `channels_rabbitmq` breaks evaluation and raises an exception.
The `channels_redis` works and sends a message fine.

Such a way looks like the `channels_redis` backend not only declares but really supports "working that way".

In order to be useful, the `channels_rabbitmq` should also support working in this mode IMHO.

I think it should monitor the connection state on calls, probably catching connection error and reconnecting again if the connection is broken. Such a solution will avoid a need to force close connection on the loop closing, as well as a need for the separate thread supporting connection to the AMQP broker.

**Comment by adamhooper** [comment]
Could you please share the exception here? I would expect channels_rabbitmq to reconnect on error as you describe. I'm curious whether this error is a bug or a behavior we can't support.

You may be able to avoid this disconnect by changing the heartbeat timeout in your RabbitMQ config.

What evidence can you share to show that channels_redis "supports" this usage? To me, "supports" means the team is willing to fix problems that arise when you use it this way. I've never heard of anybody using it this way.

**Comment by nnseva** [comment]
> Could you please share the exception here?

```
In [13]: loop.run_until_complete(get_channel_layer().group_send('simulation_event', {'type':'simulation_event', 'data':{'hello!':'world4'}}))                                                                                                

In [14]: from django.utils import timezone                                                                                                                                                                                                   

In [15]: timezone.now()                                                                                                                                                                                                                      
Out[15]: datetime.datetime(2019, 6, 13, 16, 1, 11, 255167, tzinfo=<UTC>)

In [16]: timezone.now()                                                                                                                                                                                                                      
Out[16]: datetime.datetime(2019, 6, 13, 16, 4, 26, 258384, tzinfo=<UTC>)

In [17]: loop.run_until_complete(get_channel_layer().group_send('simulation_event', {'type':'simulation_event', 'data':{'hello!':'world4'}}))                                                                                                
Connection lost exc=ConnectionResetError(104, 'Connection reset by peer')
---------------------------------------------------------------------------
ConnectionResetError                      Traceback (most recent call last)
<ipython-input-17-3124a45d5507> in <module>
----> 1 loop.run_until_complete(get_channel_layer().group_send('simulation_event', {'type':'simulation_event', 'data':{'hello!':'world4'}}))

/usr/local/lib/python3.6/asyncio/base_events.py in run_until_complete(self, future)
    482             raise RuntimeError('Event loop stopped before Future completed.')
    483 
--> 484         return future.result()
    485 
    486     def stop(self):

~/swat/py3.6/lib/python3.6/site-packages/channels_rabbitmq/core.py in group_send(self, group, message)
    214 
    215         connection = self._get_connection_for_loop()
--> 216         await connection.group_send(group, message)

~/swat/py3.6/lib/python3.6/site-packages/channels_rabbitmq/connection.py in inner(self, *args, **kwargs)
    270             raise ChannelClosed
    271 
--> 272         return await fn(self, self._channel, *args, **kwargs)
    273 
    274     return inner

~/swat/py3.6/lib/python3.6/site-packages/channels_rabbitmq/connection.py in group_send(self, channel, group, message)
    716 
    717         try:
--> 718             await channel.publish(message, self.groups_exchange, routing_key=group)
    719         except PublishFailed:
    720             # The Channels protocol has no way of reporting this error.

~/swat/py3.6/lib/python3.6/site-packages/aioamqp/channel.py in publish(self, payload, exchange_name, routing_key, properties, mandatory, immediate)
    701             await self._write_frame(self.channel_id, content_request, drain=False)
    702 
--> 703         await self.protocol._drain()
    704 
    705         if self.publisher_confirms:

~/swat/py3.6/lib/python3.6/site-packages/aioamqp/protocol.py in _drain(self)
    148             # http://bugs.python.org/issue29930. Remove this lock when no
    149             # version of Python where this bugs exists is supported anymore.
--> 150             await self._stream_writer.drain()
    151 
    152     async def _write_frame(self, channel_id, request, drain=True):

/usr/local/lib/python3.6/asyncio/streams.py in drain(self)
    337                 # would not see an error when the socket is closed.
    338                 yield
--> 339         yield from self._protocol._drain_helper()
    340 
    341 

/usr/local/lib/python3.6/asyncio/streams.py in _drain_helper(self)
    208     def _drain_helper(self):
    209         if self._connection_lost:
--> 210             raise ConnectionResetError('Connection lost')
    211         if not self._paused:
    212             return
```

**Comment by nnseva** [comment]
> What evidence can you share to show that channels_redis "supports" this usage?

For me, the best evidence is the experience. I am using channels, setting up a channels_redis and have no problem with the connection.

I don't care about a connection at all. Even more, I don't care about an asynchronous loop. I can use channels in asynchronous style in the asynchronous code, as well as in synchronous style in the synchronous code of the HTTP worker, or even in the plain synchronous script code, or in Celery task. It doesn't matter for me. Messages are sending, productivity is enough (I see some problems not dependent on sync/async code, it doesn't matter).

That's the evidence.

Now, for some reason, I would like to replace the Redis queue to the RabbitMQ queue. I am setting up a channels_rabbitmq and having a problem. It appears that I should care about RabbitMQ connection in the code. It's not what I expected - I am using channels and expecting that channels backend solves all implementation-specific details itself such a way, that all of them are hidden and transparent for me.

**Comment by adamhooper** [comment]
@nnseva Thanks for the stack trace. It looks like `channels_rabbitmq` is behaving correctly.

The exception means: RabbitMQ closed your TCP connection _while you were sending the message_.

In actuality, RabbitMQ closed the connection _before_ you sent the message. But you stopped your event loop, so `channels_rabbitmq` wasn't able to handle that connection problem. `channels_rabbitmq`'s only options at this point are: 1) silently ignore the message; or 2) raise an exception. Raising is better.

**Terrible-idea warning** You should be able to re-send the message after this exception -- now that `channels_rabbitmq` is about to enter a "reconnecting" state. After your message fails to send, `channels_rabbitmq` will reset its connection and start reconnecting. So you can `await asyncio.sleep(0.01)` (to make sure it resets the connection) and then retry the send using the exact same line as previously. Your retry will wait for reconnect and then succeed.

There's absolutely no way this retry logic enters the `channels_rabbitmq` codebase, ever. It violates the [Channels at-most-once delivery requirement](https://channels.readthedocs.io/en/latest/channel_layer_spec.html#channels). Since `channels_rabbitmq` thinks the connection dropped _during its send_, it can't know whether RabbitMQ received the message. Therefore, according to spec, the channel layer must not re-send.

All that to say:

* `channels_redis` should do the same thing if Redis drops the connection.
* You can simulate your (similarly doomed) Redis behavior by increasing the RabbitMQ [heartbeat interval](https://www.rabbitmq.com/heartbeats.html).
* You're investing a lot of time in coding something broken. Why not invest that time in something that'll work? I provided five alternate options; all of them will cost you less time than doubling down on this doomed `loop.run_until_complete()` approach.

I understand you feel `channels_rabbitmq` should make your life simpler. But the only way to make `channels_rabbitmq` work in your (dysfunctional) case would be to break it in other people's (functional) cases. My job is to avoid doing that -- and to clearly explain why.

**Comment by adamhooper** [comment]
To be clear: the problems you're having are problems with the interaction between Celery and Django Channels. Celery doesn't allow you to run threads or asyncio code; Channels doesn't work without asyncio.

In my experience, it's simplest to replace Celery.

Channels actually comes with a Celery _replacement_ called "Normal Channels". `channels_rabbitmq` doesn't support it yet; we've discussed that in #11. Personally, I'm not tempted to code "Normal Channels" because we wouldn't use it: that Channels feature is barely documented -- or, I imagine, maintained. But I've explained in #11 what is required, and I'd merge a pull request.

Here at Workbench, we replaced Celery with our own work queue, using a different rabbitmq connection (mostly copy/pasted from this project).

* [Connection](https://github.com/CJWorkbench/cjworkbench/blob/d218ce53ff07dbe22562279829f74c1a12dde369/cjworkbench/rabbitmq.py)
* [Django callers](https://github.com/CJWorkbench/cjworkbench/blob/d218ce53ff07dbe22562279829f74c1a12dde369/server/rabbitmq.py)
* [Workers](https://github.com/CJWorkbench/cjworkbench/blob/d218ce53ff07dbe22562279829f74c1a12dde369/worker/main.py) (usage: `asyncio.run(main_loop())`)

**Comment by nnseva** [comment]
Hi @adamhooper,

I've checked configuration files and agree, that the Redis looks like not breaking connection on a stalled client at all in the default configuration.

To investigate clear what happens with channels backends on connection closing, I've changed the experiment technique and restarted the Redis server instance after sending the first message. I also was using the RabbitMQ management console to ensure a broken connection in the case of RabbitMQ backend.

You right, both backends
- (1) generate an exception on broken connection if `run_until_complete` is used
- (2) restore a connection and send the second packet successfully
- (3) send a packet successfully without exception if `async_to_sync` is used

It absolutely satisfies me.

I think that such a set of options allows any use of channels (with both backends) in the asynchronous as well as the synchronous environment, and have predictable results depending on our needs and choice.

If we prefer to make minimal changes in the code and don't need productivity, we can use `async_to_sync` conversion. It guarantees that the whole necessary transmission circle - connection, sending payload, ensuring sending result and probably disconnect - happens inside one synchronous call like HTTP request using `urllib`. It looks like the most compatible and easy way to embed channels (to send some events to really asynchronous code like WebSockets application) into the synchronous code with minimal changes, _independently on the backend_.

If we would like to have a more productive code but won't change the environment, we can use `run_until_complete` call. It allows to catch a broken connection exception and resend a message if it is necessary.

If we are ready to change the whole environment to achieve the most stable **and** productive code, we can use something like Workbench - asynchronous-friendly task queue.

> You're investing a lot of time in coding something broken. Why not invest that time in something that'll work?

I've made something that works at the start. It is an `asyn_to_sync` call (you've criticized me but it really works). I was trying to invent something to more productive code. It looks like `run_until_complete` works almost as I've expected, and doesn't need global changes. Thank you for the explanation about the Channels at-most-once delivery requirement, it really prevents automatic resending a message at the reconnect and moves connection exception handling out of the backend code to my application.

Is it available for me to change the environment from Celery to Workbench? Probably yes, but it may require a lot of the legacy code migration. It is possible but requires much more time than I've already spent I'm afraid.

Thank you for the discussion and detailed explanations.
